src/main/scala/rx/lang/scala/observables/AsyncOnSubscribe.scala

```diff
@@ -0,0 +1,120 @@
+    new AsyncOnSubscribeImpl[Unit, T](() => (), (_, req, obs) => next(req, obs), _ => onUnsubscribe()).toObservable
+
+  /**
+    * Generates an `Observable` that calls the provided `next` function to generate observables with at most the requested amount of items.
+    *
+    * @tparam T the type of the generated values
+    * @tparam S the type of the associated state with each Subscriber
+    * @param generator generates the initial state value
+    * @param next produces an observable with at most the requested amount of items for the stream
+    * @param onUnsubscribe clean up behavior
+    * @return An observable that emits data downstream in a protocol compatible with back-pressure.
+    */
+  @Experimental
+  def apply[S, T](generator: () => S)(next: (S, Long) => (Notification[Observable[_ <: T]], S), onUnsubscribe: S => Unit = (_:S) => ()): Observable[T] = {
+    val nextF: (S, Long, Observer[Observable[_ <: T]]) => S =
+      (state, requested, observer) => next(state, requested) match {
+        case (notification, nextState) =>
+          notification.accept(observer)
+          nextState
+      }
+    new AsyncOnSubscribeImpl[S, T](generator, nextF, onUnsubscribe).toObservable
+  }
+
+  /**
+    * Generates an `Observable` that calls the provided `next` function to generate observables with at most the requested amount of items.
+    *
+    * @tparam T the type of the generated values
+    * @tparam S the type of the associated state with each Subscriber
+    * @param generator generates the state value
+    * @param next produces an observable with at most the requested amount of items for the stream
+    * @param onUnsubscribe clean up behavior
+    * @return An observable that emits data downstream in a protocol compatible with back-pressure.
+    */
+  @Experimental
+  def applySingleState[S, T](generator: () => S)(next: (S, Long) => Notification[Observable[_ <: T]], onUnsubscribe: S => Unit = (_:S) => ()): Observable[T] =
+    apply[S, T](generator)((s,r) => (next(s,r), s), onUnsubscribe)
+
+  /**
+    * Generates an `Observable` that calls the provided `next` function to generate observables with at most the requested amount of items.
+    *
+    * @tparam T the type of the generated values
+    * @param next produces an observable with at most the requested amount of items for the stream
+    * @param onUnsubscribe clean up behavior
+    * @return An observable that emits data downstream in a protocol compatible with back-pressure.
+    */
+  @Experimental
+  def applyStateless[T](next: Long => Notification[Observable[_ <: T]], onUnsubscribe: () => Unit = () => ()) =
+    apply[Unit, T](() => ())((_,r) => (next(r), ()), _ => onUnsubscribe())
+
+  private[scala] class AsyncOnSubscribeImpl[S,T](val generatorF: () => S, val nextF: (S, Long, Observer[Observable[_ <: T]]) => S, val onUnsubscribeF: S => Unit) {
+    import rx.lang.scala.JavaConversions._
```

Comment:
Since there were plans to deprecate `JavaConversions` (see [comment by @zsxwing](https://github.com/ReactiveX/RxScala/pull/207#issuecomment-249042192)), maybe you should pro-actively use `JavaConverters` instead. (not sure what the status of this plan currently is, though!)